apps/rush-lib/src/pluginFramework/RushSessionForPlugin.ts

```diff
@@ -0,0 +1,118 @@
+import { RushConstants } from '../logic/RushConstants';
+import { ContributionPoint, IContributeAPIForBuildCacheProvider } from './ContributionPoint';
+import { IRushPluginManifest } from './PluginLoader/PluginLoaderBase';
+import { RushLifecycleHooks } from './RushLifeCycle';
+import { RushSession } from './RushSession';
+
+interface IContributeCallInfo {
+  expectedFunctionNameSet: Set<string>;
+  calledFunctionNameSet: Set<string>;
+}
+
+/**
+ * @beta
+ *
+ * When a plugin specifies contribution points, it means the plugin must call
+ * all of the according APIs in rushSession once before initialized. This
+ * ensure the semantics of the contribution points which the plugin specifies.
+ */
+export class RushSessionForPlugin extends RushSession {
+  private _isOutOfScope: boolean = false;
+  private _contributeInfo: Map<ContributionPoint, IContributeCallInfo> = new Map<
+    ContributionPoint,
+    IContributeCallInfo
+  >();
+  private _pluginManifest: IRushPluginManifest;
+  public constructor(internalRushSession: RushSession, pluginManifest: IRushPluginManifest) {
+    super(internalRushSession.options);
+
+    // sync internal variables with internalRushSession
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (this as any)._cloudBuildCacheProviderFactories =
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (internalRushSession as any)._cloudBuildCacheProviderFactories;
+
+    this._pluginManifest = pluginManifest;
+
+    if (!Array.isArray(pluginManifest.contributes)) {
+      return;
+    }
+
+    const hookForPlugin: RushLifecycleHooks = new RushLifecycleHooks();
+
+    // Plugins specified contribution point should be initialize lazily.
+    this.hooks.initialize = hookForPlugin.initialize;
```

Comment:
As of #3252, `initialize` is no longer the only hook for plugins, there are now specific hooks for phased/global command execution, with the purpose of allowing the plugin access to the executing command (e.g. the PhasedScriptAction or GlobalScriptAction instance).